ioredis; atomic locking with redlock; support Node 6.x [Fixes #185, Fixes #370]

doublerebel opened the pull request:
Hello,

This PR changes the redis dependency from **node_redis** to **ioredis**, fixing #185 (and completing PR #188).  **ioredis** uses promises "all-the-way-down", exposing errors in the test suite.  These tests are fixed and passing, including the race condition described in #370.

There are several commits here, to demonstrate the bugs and fixes made to the test suite.  I am happy to squash this down, but right now each commit contains a description of the issues solved with the updated tests.

As part of solving #370, [**node-redlock**](https://github.com/mike-marcacci/node-redlock) is now used to implement atomic locking with a guaranteed owner.  **node-redlock** is a promise-based lib which keeps track of its own secret token/value for each job/resource, which makes node-redlock valuable whether or not it's used with redis clustering.

Thanks for your work, looking forward to your feedback.

manast commented on lib/queue.js:
```diff
@@ -147,11 +147,16 @@ var Queue = function Queue(name, redisPort, redisHost, redisOptions){
+  });
+
+  this._initializing = Promise.join.apply(null, initializers)
+  .then(function(){
```
here we could also use ```Promise.all(initializers)``` for more succinct code.

manast commented on lib/queue.js:
```diff
@@ -147,11 +147,16 @@ var Queue = function Queue(name, redisPort, redisHost, redisOptions){
-    this.eclient.select(redisDB)
-  ).then(function(){
+  var initializers = [];
+  [this.client, this.bclient, this.eclient].forEach(function (client) {
```
use .map instead of forEach for a shorter and more semantically correct code.

manast commented on test/test_cluster.js:
```diff
@@ -44,18 +44,21 @@ function workerMessageHandlerWrapper(message) {
+    }
+  });
+
   var queue;
```
great improvement!

manast commented on lib/queue.js:
```diff
@@ -257,7 +257,9 @@ Queue.prototype.distEmit = function(){
+  var promise = Disturbed.prototype.on.apply(this, args);
+  var _this = this;
+  promise.catch(function(err){ _this.emit('error', err); });
   return this;
```
I changed the API in disturbed so that it will handle local events as well which will break this. I'll fix it asap.

manast commented on lib/scripts.js:
```diff
@@ -304,6 +304,7 @@ var scripts = {
    */
   takeLock: function(queue, job, token, renew){
+    console.log('attempting lock for job id #', job.jobId)
     var lockCall;
```
we may need a debug flag to switch on/off this debug logs.

doublerebel replied:
This log is removed in a later commit, I'll make sure I have no such console.* statements in the final PR.

manast commented on lib/scripts.js:
```diff
@@ -185,7 +180,9 @@ var scripts = {
+    return job.takeLock(null, !!job.lock)
+    .then(function(lock) { return lock ? execScript.apply(scripts, args) : -2; }, function(err) { if (!(err instanceof Redlock.LockError)) { throw err; } })
+    .then(function(result){
       switch (result){
```
this could be formatted into several lines or refactored into a separate function

doublerebel replied:
You're right, I'll clean this up.

manast commented on lib/scripts.js:
```diff
@@ -303,59 +302,28 @@ var scripts = {
+      throw new Error('Unable to renew nonexisting lock');
+    } else if (renew) {
+      return lock.extend(queue.LOCK_RENEW_TIME);
+    } else if (lock) {
```
I think the if here is not needed, lock must be defined if we reached this point.

doublerebel replied:
In this case, if the lock is defined we can return null early and not need to call redlock.

manast replied:
yes, I understand that, what I mean is that lock will always defined in line 311, so you don't need the if, the else is enough.

doublerebel replied:
I definitely don't need else after return or throw.  Updated in this latest push.

manast commented on lib/scripts.js:
```diff
@@ -303,59 +302,28 @@ var scripts = {
+      return lock.extend(queue.LOCK_RENEW_TIME);
+    } else if (lock) {
+      return new Promise(function(resolve) {
+        resolve(null);
```
just use ```return Promise.resolve(null);```

manast commented on lib/scripts.js:
```diff
@@ -315,7 +315,8 @@ var scripts = {
-    return redlock.lock(job.lockKey(), queue.LOCK_RENEW_TIME);
+    return redlock.lock(job.lockKey(), queue.LOCK_RENEW_TIME)
+    .then(job.queue.client.hset(job.queue.toKey(job.jobId), "lockAcquired", 1));
   },
```
the problem here is that getting the lock and setting the lockAcquired property is not an atomic operation anymore...

manast commented on lib/scripts.js:
```diff
@@ -314,9 +314,31 @@ var scripts = {
+      var lockAcquired = 'and redis.call("HSET", "' + queue.toKey(job.jobId) + '", "lockAcquired", "1")';
+      var success = 'then return 1 else return 0 end';
+      var opts = {
+        lockScript: function(lockScript) {
```
ok, so here we can actually "monkey patch" the redlock code. I was actually wondering how we could solve the issue with ensuring that the job is in the active list while getting a redlock atomically. But how robust is this when new versions of redlock are released? Isn't there a risk that our code would stop working?

manast commented:
great PR. A Lot of good stuff all over the place. Please check my comments. I am sure @bradvogel will also give you some valuable feedback.

doublerebel commented on lib/queue.js:
```diff
@@ -581,26 +577,30 @@ Queue.prototype.moveUnlockedJobsToWait = function(){
-          return null;
-        });
+  return scripts.moveUnlockedJobsToWait(this).then(function(responses){
+    var handleFailedJobs = responses[0].map(function(jobId){
```
Ooh, good catch.

doublerebel replied:
I tried this, and it seems that spread must have an array, but in bull sometimes scripts.moveUnlockedJobsToWait returns undefined:
```js
 Cluster
Worker spawned: # 1
Worker spawned: # 2
Worker spawned: # 3
Worker spawned: # 4
Worker spawned: # 5
Worker spawned: # 6
Worker spawned: # 7
Failed to handle unlocked job in active: TypeError: Cannot read property 'map' of undefined
    at /home/charles/source/bull/lib/queue.js:576:40
    at tryCatcher (/home/charles/source/bull/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/charles/source/bull/node_modules/bluebird/js/release/promise.js:507:35)
    at Promise._settlePromise (/home/charles/source/bull/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/home/charles/source/bull/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/home/charles/source/bull/node_modules/bluebird/js/release/promise.js:691:18)
```

doublerebel commented:
Thanks @manast.  I really appreciate all the Promise API improvements, I'm not as familiar with it.  I'll rebase this PR, clean it up and repush.

doublerebel commented:
I rebased, and cleaned up all the commit based on your feedback.  The previous set of commits is saved at [nextorigin/bull/commits/issue-185-dirty](https://github.com/nextorigin/bull/commits/issue-185-dirty).

Now there are two major commits: the ioredis refactor, and the redlock refactor, almost all the rest are test improvements.  I get all tests passing on every run here, so I hope that is reflected shortly in travis 😃

doublerebel commented:
With the lock and promise improvements, this PR might fix #333 as well.  Leaving this note for future reference.

bradvogel commented on package.json:
```diff
@@ -20,10 +20,11 @@
     "mocha": "^3.1.2",
     "node-uuid": "^1.4.7",
-    "redis": "^2.6.3",
+    "redlock": "nextorigin/node-redlock",
```
I don't think Bull should have this dependency. Please try to upstream your fix into node-redlock.

bradvogel commented on lib/scripts.js:
```diff
@@ -305,93 +311,51 @@ var scripts = {
-      lockCall = 'redis.call("SET", KEYS[1], ARGV[1], "PX", ARGV[2])';
+    var redlock;
+    if (ensureActive) {
+      var isJobInList = this._isJobInList.replace(/KEYS\[1\]/, '"' + job.queue.toKey('active') + '"')
```
This is too fragile to be running string replacement. Please make `_isJobInList` a function that generates the appropriate string.

doublerebel replied:
You're right, I'll clean that up.

bradvogel commented:
What about all the scripts that look for `:lock` in the key such as `moveUnlockedJobsToWait`? Won't these break?

bradvogel commented on lib/scripts.js:
```diff
@@ -305,93 +311,51 @@ var scripts = {
+    if (renew && !lock) {
+      throw new Error('Unable to renew nonexisting lock');
+    } else if (renew) {
+      return lock.extend(queue.LOCK_RENEW_TIME);
```
The lock renewer runs on `_this.LOCK_RENEW_TIME / 2` interval, so this will keep extending it out farther and farther on each renewal. So if it gets stalled, it'll potentially be locked for many minutes. Is there a way to extend it to the absolute time `(currentTime + queue.LOCK_RENEW_TIME)`?

doublerebel replied:
The [redlock extend method uses PEXPIRE](https://github.com/nextorigin/node-redlock/blob/master/redlock.js#L16), and should be functionally equivalent to the SET ... PX method bull uses now.  It only sets the expire from the current time, it does not add to the existing TTL, so it's typical redis expire behavior.

bradvogel replied:
Oh yes, you're right. Filed https://github.com/mike-marcacci/node-redlock/issues/26 to clarify their docs.

doublerebel commented:
@bradvogel they actually work exactly the same, redlock's method is very similar to bull's method and it works with `moveUnlockedJobsToWait` as a drop-in replacement.

We can see `moveUnlockedJobsToWait` working in some of the tests, the long process time is the stalled script running twice to move the jobs here:
https://travis-ci.org/OptimalBits/bull/jobs/175763931#L823

Also in my previous runs where I was logging heavily, I logged that process in particular so I could ensure it was working, the `moveUnlockedJobsToWait` calls are logged here:
https://travis-ci.org/nextorigin/bull/jobs/175597819#L1652

bradvogel commented:
Ok, I see. I verified moveUnlockedJobsToWait still works locally. Thanks!

doublerebel commented:
@bradvogel For sure, I appreciate the push for code quality.  I've rewritten `._isJobInList` in 3a41230 and re-pushed.

bradvogel commented:
btw, I'm not completely familiar with Redis cluster, but a potential hazard that might cause issues with cluster is how we concatenate job keys in some scripts: https://github.com/nextorigin/bull/blob/issue-185/lib/scripts.js#L439. It's my understanding that if keys aren't passed in to scripts via the `KEYS` argument, then there's no guarantee that they will run on the right cluster (http://redis.io/commands/EVAL). So we might need to rethink how we run these scripts. Anyways, outside the scope of this PR.

doublerebel commented:
@bradvogel good call, I'll make a note of that.  Perhaps we can ask for some user stories for bull+redis cluster usage from those issues.  I run a lot of jobs, and have been using it with Sentinel+Redishappy+Consul for failover, but not a redis cluster per se.

bradvogel commented on lib/queue.js:
```diff
@@ -615,17 +637,14 @@ Queue.prototype.processJob = function(job){
-    // to do this on subsequent calls.
-    return job.takeLock(_this.token, renew, !renew).then(function(locked){
-      if(locked){
+    return job.takeLock(renew, true).then(function(lock){
```
Mismerge. The second parameter should be `!renew`. Also include comment

doublerebel replied:
Here [`takeLock` returns early](https://github.com/nextorigin/bull/blob/5c1e9d9b5e8e480050b558c197e37a48210b924c/lib/scripts.js#L327) and relies on redlock to extend the lock, `ensureActive` is only considered for new locks.  So I don't know if insinuating the second parameter matters on a renew is necessary?

manast commented:
please let me know if everybody is happy so that I can manually merge this in the morning swedish time.

albertjan commented:
Looks brilliant, does this introduce a new locking mechanism? If so is there some docs on that because I'll need to update fsprojects/oxen to stay compatible with bull. :)

bradvogel commented:
I'll give it one more thorough review in a few hours.

@manast do you think this should be released as 2.0? While technically none of the APIs documented in the readme have changed with this PR, anyone using undocumented `createClient` (https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L96) will now need to create an ioredis client. What do you think?

bradvogel commented:
Ran a few more tests and this looks good to me. There's a merge conflict though.

doublerebel commented:
Must be from today's new commits on master.  I can rebase again first thing tomorrow.

On November 14, 2016 9:23:05 PM PST, Brad Vogel <notifications@github.com> wrote:
>Ran a few more tests and this looks good to me. There's a merge
>conflict though.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/OptimalBits/bull/pull/379#issuecomment-260550483

manast commented:
I guess it is safer to release it as 2.0, to avoid the risk of breaking users code. I can write a notice explaining that the changes are mostly internal but large enough to motivate a major release.

bradvogel commented on lib/scripts.js:
```diff
@@ -305,93 +316,51 @@ var scripts = {
+      redlock = new Redlock(queue.clients, _.extend(opts, queue.redlock));
     } else {
-      lockCall = 'redis.call("SET", KEYS[1], ARGV[1], "PX", ARGV[2], "NX")';
+      redlock = new Redlock(queue.clients, queue.redlock);
```
I personally think it's cleaner to monkey patch Redlock (eg `redlock.lockScript = ...`) for now until PR https://github.com/mike-marcacci/node-redlock/pull/25 is resolved. To mitigate against Redlock changing, just lock the version in the package.json (so `"redlock": "2.0.2"`).

doublerebel replied:
Unfortunately it's a private variable without my patch.  I've petitioned for an answer from the dev, I can fork it if necessary, I maintain forks of other packages too.

bradvogel replied:
Yeah, that's what I mean - overwrite the private variable. Lock the version down in package.json (eg `"redlock": "2.0.2"`) to make sure that variable won't change on us. Then in code:

```
var redlock = new Redlock(..)
redlock.lockScript = redlock.lockScript.replace(...)
```

This is much safer because it still uses the real `redlock` module, and a locked version that can't update. It's clearer in package.json so other devs can see what version we're using. You can even add a comment in package.json (http://stackoverflow.com/questions/14221579/how-do-i-add-comments-to-package-json-for-npm-install) to indicate why it's locked to a specific version.

Pointing this to `nextorigin/node-redlock` is dangerous because if you or anyone on your team accidentally pushed to that master branch, it'll break Bull for **all** users. I trust that you wouldn't do that, but it's unnecessary risk.

cc @manast

manast commented:
@albertjan it uses a different locking mechanism (redlock) that is supposed to work in a distributed environment (needed when running redis cluster). I wondering though if we are going to announce that bull works with redis cluster, I have not made any tests myself, and as mentioned before, our LUA scripts may not work well in such an environment.

doublerebel commented:
@albertjan 
Redlock algo is almost exactly like bull's current lock logic, it's the technique described in [redis distlock](http://redis.io/topics/distlock).  node-redlock keeps track of its own per-job secret, so the javascript process owns the lock and checks on renew just the same.  To extend the lock, node-redlock uses `PEXPIRE` where bull uses `SET ... PX`, but that is just syntax.  It's convenient to have the lock packaged in a promise-based, tested codebase.

Part of the reason to implement ioredis is to start exploring redis sentinel and cluster support, however as described in [redis distlock](http://redis.io/topics/distlock#why-failover-based-implementations-are-not-enough), bull will lose lock guarantees in a failover scenario unless redlock is implemented.  node-redlock manages those cluster guarantees as well.

There are some .NET libraries for redlock [listed in redis distlock](http://redis.io/topics/distlock#implementations).  Bull is now doing up to 3 operations in one block to get a lock:
* check to see if job is in active list
* set the lock
* set the lockAcquired counter

Unfortunately it looks like none of the .NET libs have the lock script as a publicly accessible or modify-able variable, I have a [PR open for node-redlock](https://github.com/mike-marcacci/node-redlock/issues/25) for that functionality.

doublerebel commented:
Rebased on master and re-pushed.  Fingers crossed for node-redlock#25.

doublerebel commented:
@bradvogel I totally agree about pulling packages from Github, I would have released a new package like I did with stream-combiner2-withopts.  The lockscript was private like trapped in a closure private, but thankfully the PR was merged.  I swear 90% of the time I have to maintain a forked package, it's just because some option is trapped in a closure.

The only change in this latest push is to update redlock to `^2.1.0` in package.json.  That should complete it unless you guys caught anything else!
